### PR TITLE
Add scalar_loss for LossInfo

### PR DESCRIPTION
**Pull request by emailweixu**
Also
1. updated ICMAlorithm to support multiple hidden layers.
2. minor refactoring of ActorCriticAlgorithm so that it's easier to add new sub-algorithm in the future.
3. improved icm_mountain_car.gin so it can train within an hour

### CODE REVIEW

File: alf/algorithms/icm_algorithm.py [hunk 1]
```diff
@@ -134,9 +163,15 @@ def train_step(self, inputs, state, calc_intrinsic_reward=True):
...
-                    forward_loss=forward_loss, inverse_loss=inverse_loss)))
+            info=ICMInfo(
+                reward=intrinsic_reward,
+                loss=LossInfo(
+                    loss=forward_loss + inverse_loss,
+                    extra=dict(
+                        forward_loss=forward_loss,
+                        inverse_loss=inverse_loss))))
+
+    def calc_loss(self, info: ICMInfo):
```
Comment (hnyu) [inline alf/algorithms/icm_algorithm.py]: should avoid masking out across-episode transitions by converting the loss to scalar. Maybe define a wrapper called mask_invariant_loss()

File: alf/algorithms/rnd_algorithm.py [hunk 2]
```diff
@@ -102,4 +104,7 @@ def train_step(self, inputs, state):
         return AlgorithmStep(
             outputs=intrinsic_reward,
```
Comment (hnyu) [inline alf/algorithms/rnd_algorithm.py]: outputs should be ()

File: alf/utils/summary_utils.py [hunk 3]
```diff
@@ -221,3 +221,32 @@ def summarize_action_dist(action_distributions,
...
+def add_mean_hist_summary(name, value):
+    """Generate mean and histogram summary of value.
+
+    Args:
+        name (str): name of the summary
+        value (Tensor): tensor to be summarized
+    Returns:
+        None
+    """
+    tf.summary.histogram(name + "/value", value)
```
Comment (hnyu) [inline alf/utils/summary_utils.py]: This function can call the one below.

File: alf/algorithms/icm_algorithm.py (hunk 1 above)
Comment (emailweixu) [reply alf/algorithms/icm_algorithm.py]: added reduce_mean

**Pull request merged**